extensions/eclipse-che-theia-plugin-remote/src/node/webviews-content-aware.ts

```diff
@@ -0,0 +1,104 @@
+
+import { Plugin } from '@theia/plugin-ext/src/common/plugin-api-rpc';
+import { Uri } from '@theia/plugin';
+import { overrideUri } from './che-content-aware-utils';
+
+export class RemoteWebview extends WebviewImpl {}
+
+export class WebviewsContentAware {
+  static makeWebviewsContentAware(webviewExt: WebviewsExtImpl): void {
+    const webviewsContentAware = new WebviewsContentAware();
+    webviewsContentAware.overrideVSCodeResourceScheme(webviewExt);
+  }
+
+  overrideVSCodeResourceScheme(webviewExt: WebviewsExtImpl): void {
+    this.rebind$createWebview(webviewExt);
+  }
+
+  // Modify WebviewOptions#localResourceRoots by setting remote side car scheme instead of default file
+  // during webview panel create step. This method activates only when plugin call theia.createWebview
+  // method from remote sidecar. Normally from theia sidecar this method should not be executed.
+  //
+  // localResourceRoots provides paths where extension hosts own resources, styles, scripts, fonts, etc.
+  private rebind$createWebview(webviewExt: WebviewsExtImpl): void {
+    const original$createWebview = webviewExt.createWebview.bind(webviewExt);
+    webviewExt.createWebview = (
+      viewType: string,
+      title: string,
+      showOptions: theia.ViewColumn | theia.WebviewPanelShowOptions,
+      options: theia.WebviewPanelOptions & theia.WebviewOptions,
+      plugin: Plugin
+    ) => {
+      const webviewPanel: theia.WebviewPanel = original$createWebview(
+        viewType,
+        title,
+        showOptions,
+        options.localResourceRoots
+          ? ({
+              enableFindWidget: options.enableFindWidget,
+              retainContextWhenHidden: options.retainContextWhenHidden,
+              enableScripts: options.enableScripts,
+              enableCommandUris: options.enableCommandUris,
+              localResourceRoots: (() => options.localResourceRoots.map(root => overrideUri(root)))(),
+              portMapping: options.portMapping,
+            } as theia.WebviewPanelOptions & theia.WebviewOptions)
+          : options,
+        plugin
+      );
+
+      this.rebind$asWebviewUri(webviewPanel.webview);
+      this.rebind$_htmlSetter(webviewPanel.webview);
+
+      return webviewPanel;
+    };
+  }
+
+  // Method theia.Webview#asWebviewUri is being called from theia container.
+  // In default flow method returns the path like: /webview/theia-resource/file:///path/to/directory
+  // with https scheme and authority
+  private rebind$asWebviewUri(webview: theia.Webview): void {
+    const original$asWebviewUri = webview.asWebviewUri.bind(webview);
+    webview.asWebviewUri = (resource: Uri) => original$asWebviewUri(overrideUri(resource));
+  }
+
+  // Browser part perform preprocess initial html content by replacing vscode-resource:/path/to/file
+  // to https://authority/webview/theia-resource/file/path/to/file. To be able to operate with custom
```

Comment:
we can't for example just patch webview/normalizeRequestUri  https://github.com/eclipse-theia/theia/blob/fd0a8ce093273011ecf4500d673a982854af9e55/packages/plugin-ext/src/main/browser/webview/webview.ts

Reply:
I tried this solution, but it seems, that this won't work. The problem is extending the WebviewWidget won't get a proper effect, because this code is run on browser part on Theia sidecar and we need to provide the container name from environment variable on the sidecar from plugin.

Reply:
> I tried this solution, but it seems, that this won't work.

sorry @vzhukovs but it's not clear to me if
you think that this will not work OR you have tried and this didn't work
Could you please clarify it? :slightly_smiling_face:

Reply:
I've tried. The main problem here is that rebound class will be executing on Theia's container where we don't have any sidecar specific information. As far as rebound class will execute from the browser part we also don't have ability to get environment information. So it better to intercept this method's calls from server side on sidecar directly:

VS Code extension (sidecar server side) > Method intercept and link modification (sidecar server side) > Plugin Ext (sidecar server side > theia browser side) > Theia (theia browser side).

Reply:
we can't only rebind only in sidecar usecase ?

Reply:
From what I saw, I couldn't get this worked